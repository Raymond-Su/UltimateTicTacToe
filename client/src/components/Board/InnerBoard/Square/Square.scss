.cell {
  width: calc(14px + 2.5vw);
  height: calc(14px + 2.5vw);
  font-size: calc(11px + 1.1vw);
  font-weight: 300;

  @media (max-width: 992px) {
    font-weight: 400;
  }

  @media (max-width: 768px) {
    font-weight: 500;
  }

  &.W,
  &.SW,
  &.S,
  &.SE {
    border-top: 2px solid gray;
  }

  &.NW,
  &.N,
  &.NE,
  &.W,
  &.E {
    border-bottom: 2px solid gray;
  }

  &.N,
  &.NE,
  &.E,
  &.S,
  &.SE {
    border-left: 2px solid gray;
  }

  &.NW,
  &.N,
  &.W,
  &.SW,
  &.S {
    border-right: 2px solid gray;
  }

  &.enabled:hover {
    cursor: pointer;
    background-color: silver;
  }
}
