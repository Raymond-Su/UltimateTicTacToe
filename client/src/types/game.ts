// Enums
export enum Opponent {
  Player = 'player',
  AI = 'ai'
}

export enum AIDifficulty {
  Easy,
  Medium,
  Hard,
  Expert,
  Grandmaster,
  Impossible
}

export enum Player {
  Cross = 0,
  Circle = 1,
  Spectator = 2
}

export enum TileValue {
  Cross = 0,
  Circle = 1,
  Empty = 2,
  Destroyed = 3
}

export enum Winner {
  Cross = 0,
  Circle = 1,
  None = 2,
  Draw = 3
}

// Interfaces
export interface Point {
  readonly x: number;
  readonly y: number;
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}

export interface TileInformation {
  position: Point;
  value: TileValue;
}

export interface innerBoardInformation extends TileInformation {
  tiles: SquareInformation[];
}

export interface SquareInformation extends TileInformation {
  boardPosition: Point;
}

export interface WinResult {
  isFinished: boolean;
  winningPlayer: Winner;
}
