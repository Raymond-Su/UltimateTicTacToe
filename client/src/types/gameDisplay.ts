export enum GameAction {
  MOVE = 'MOVE',
  RESTART = 'RESTART'
}
