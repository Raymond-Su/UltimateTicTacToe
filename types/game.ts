export enum Player {
  Cross = 0,
  Circle = 1,
  Spectator = 2
}

export interface GameState {
  playerCount: number;
  playerX: string | null;
  playerO: string | null;
  history: Move[];
}

export interface Point {
  readonly x: number;
  readonly y: number;
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}
