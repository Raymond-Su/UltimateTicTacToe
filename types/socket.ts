export enum SocketCore {
  CONNECTION = 'connection',
  ERROR = 'error',
  DISCONNECT = 'disconnect'
}

export enum GameSocketClientMessage {
  MAKE_MOVE = 'MAKE_MOVE'
}

export enum GameSocketServerMessage {
  MOVE_MADE = 'MOVE_MADE',
  SET_PLAYER = 'SET_PLAYER',
  CURRENT_BOARD = 'CURRENT_BOARD'
}
