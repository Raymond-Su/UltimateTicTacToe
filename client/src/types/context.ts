import TicTacToeGame from '../utils/gameEngine';
import { GameAction } from './gameDisplay';

export interface AppState {
  game: TicTacToeGame;
}

export interface StateAction {
  type: GameAction;
  payload?: any;
}
