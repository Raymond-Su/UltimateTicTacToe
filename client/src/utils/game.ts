import { Player, Point, TileValue, Winner } from '../types/game';

export const displayTileValue: Record<TileValue, string> = {
  [TileValue.Cross]: 'X',
  [TileValue.Circle]: 'O',
  [TileValue.Destroyed]: '/',
  [TileValue.Empty]: ''
};

export const displayPlayerValue: Record<Player, string> = {
  [Player.Cross]: 'X',
  [Player.Circle]: 'O',
  [Player.Spectator]: 'Spectator'
};

export const playerToWinner = (player: Player): Winner => {
  switch (player) {
    case Player.Circle:
      return Winner.Cross;
    case Player.Cross:
      return Winner.Circle;
    default:
      return Winner.None;
  }
};

export const playerToTileValue = (
  player: Winner | Player,
  isForFullInnerBoard = false
): TileValue => {
  if (isForFullInnerBoard && player === Winner.Draw) {
    return TileValue.Destroyed;
  } else if (player === Winner.Cross) {
    return TileValue.Cross;
  } else if (player === Winner.Circle) {
    return TileValue.Circle;
  } else {
    return TileValue.Empty;
  }
};

export const arePointsEqual = (point1: Point, point2: Point): boolean => {
  if (point1.x === point2.x && point1.y === point2.y) {
    return true;
  }
  return false;
};
