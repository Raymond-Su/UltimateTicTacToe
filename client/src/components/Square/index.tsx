import './Square.scss';

import classNames from 'classnames';
import React, { FC } from 'react';

import { TileValue } from '../../types/game';
import { displayTileValue } from '../../utils/gameUtils';

interface SquareProps {
  isActive?: boolean;
  notationPosition?: string;
  tileValue: TileValue;
  makeMove: () => void;
}

const Square: FC<SquareProps> = ({
  isActive = false,
  notationPosition = '',
  tileValue,
  makeMove
}: SquareProps) => {
  return (
    <td
      className={classNames(`cell ${notationPosition}`, {
        enabled: isActive && tileValue === TileValue.Empty
      })}
      onClick={makeMove}
      data-testid="game-square"
    >
      {displayTileValue[tileValue]}
    </td>
  );
};

export default Square;
