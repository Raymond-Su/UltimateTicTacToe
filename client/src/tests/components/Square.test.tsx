import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { create } from 'react-test-renderer';

import Square from '../../components/Board/InnerBoard/Square';
import { TileValue } from '../../types/game';

const mockFunction = jest.fn();

const renderOptions = {
  container: document.body.appendChild(document.createElement('tr'))
};

describe('<Square />', () => {
  test('Renders component', () => {
    const { getByTestId } = render(
      <Square makeMove={mockFunction} tileValue={TileValue.Empty} />,
      renderOptions
    );
    const square = getByTestId('game-square');
    expect(square).toHaveClass('cell');
    expect(square).toHaveTextContent('');
  });

  test('Notation class added to component', () => {
    const notationPosition = 'NW';
    const { getByTestId } = render(
      <Square
        makeMove={mockFunction}
        tileValue={TileValue.Empty}
        notationPosition={notationPosition}
      />,
      renderOptions
    );
    const square = getByTestId('game-square');
    expect(square).toHaveClass('cell', notationPosition);
  });

  test('Square is enabled, when tileValue is EMPTY and isActive is true', () => {
    const { getByTestId } = render(
      <Square
        makeMove={mockFunction}
        tileValue={TileValue.Empty}
        isActive={true}
      />,
      renderOptions
    );
    const square = getByTestId('game-square');
    expect(square).toHaveClass('cell', 'enabled');
  });

  test('Square is not enabled, when tileValue is EMPTY and isActive is false', () => {
    const { getByTestId } = render(
      <Square
        makeMove={mockFunction}
        tileValue={TileValue.Empty}
        isActive={false}
      />,
      renderOptions
    );
    const square = getByTestId('game-square');
    expect(square).toHaveClass('cell');
    expect(square).not.toHaveClass('enabled');
  });

  test('Square is not enabled when tileValue is not EMPTY and isActive is true', () => {
    const { getByTestId } = render(
      <Square
        makeMove={mockFunction}
        tileValue={TileValue.Cross}
        isActive={true}
      />,
      renderOptions
    );
    const square = getByTestId('game-square');
    expect(square).toHaveClass('cell');
    expect(square).not.toHaveClass('enabled');
  });

  test('displays X when tileValue is TileValue.Cross', () => {
    const { getByTestId } = render(
      <Square makeMove={mockFunction} tileValue={TileValue.Cross} />,
      renderOptions
    );
    const square = getByTestId('game-square');
    expect(square).toHaveTextContent('X');
  });

  test('fires madeMove function when clicked', () => {
    const { getByTestId } = render(
      <Square makeMove={mockFunction} tileValue={TileValue.Empty} />,
      renderOptions
    );
    const square = getByTestId('game-square');
    fireEvent.click(square);
    expect(mockFunction).toBeCalledTimes(1);
  });

  test('Matches the snapshot', () => {
    const square = create(
      <Square
        makeMove={mockFunction}
        tileValue={TileValue.Circle}
        isActive={false}
        notationPosition="NW"
      />
    );
    expect(square.toJSON()).toMatchSnapshot();
  });
});
