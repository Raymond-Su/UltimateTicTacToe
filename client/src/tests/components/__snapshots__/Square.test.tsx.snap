// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Square /> Matches the snapshot 1`] = `
<td
  className="cell NW"
  data-testid="game-square"
  onClick={[MockFunction]}
>
  O
</td>
`;
